File: lightbulb-recycleradapters/src/main/java/com/github/RooneyAndShadows/lightbulb/recycleradapters/implementation/CheckBoxSelectableAdapter.java
package com.github.rooneyandshadows.lightbulb.recycleradapters.implementation;

import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.ViewGroup;


import com.github.rooneyandshadows.lightbulb.recycleradapters.R;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterConfiguration;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterDataModel;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterSelectableModes;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyRecyclerAdapter;
import com.github.rooneyandshadows.lightbulb.selectableview.CheckBoxView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

@SuppressWarnings("unchecked")
public class CheckBoxSelectableAdapter<ItemType extends EasyAdapterDataModel> extends EasyRecyclerAdapter<ItemType> {

    public CheckBoxSelectableAdapter() {
        super(new EasyAdapterConfiguration<ItemType>().withSelectMode(EasyAdapterSelectableModes.SELECT_MULTIPLE));
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        CheckBoxView v = (CheckBoxView) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.layout_checkbox_button, parent, false);
        int[] padding = getItemPadding();
        if (padding != null && padding.length == 4)
            v.setPadding(padding[0], padding[1], padding[2], padding[3]);
        return new CheckBoxViewHolder(v, this);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        CheckBoxViewHolder vHolder = (CheckBoxViewHolder) holder;
        vHolder.bindItem();
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        CheckBoxViewHolder vh = (CheckBoxViewHolder) holder;
        vh.recycle();
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public String getItemName(ItemType item) {
        return item.getItemName();
    }

    protected Drawable getItemIcon(ItemType item) {
        return null;
    }

    protected Drawable getItemIconBackground(ItemType item) {
        return null;
    }

    /**
     * Defines padding for the item view.
     *
     * @return int[] {left,top,right,bottom}
     */
    protected int[] getItemPadding() {
        return null;
    }

    public class CheckBoxViewHolder extends RecyclerView.ViewHolder {
        protected CheckBoxView selectableView;
        protected ItemType item;

        CheckBoxViewHolder(CheckBoxView categoryItemBinding, EasyRecyclerAdapter<ItemType> adapter) {
            super(categoryItemBinding);
            this.selectableView = (CheckBoxView) itemView;
            selectableView.setOnCheckedListener((view, isChecked) -> selectableView.post(() -> adapter.selectItemAt(getBindingAdapterPosition(), isChecked)));
        }

        public void bindItem() {
            this.item = getItem(getBindingAdapterPosition());
            boolean isSelectedInAdapter = isItemSelected(item);
            if (selectableView.isChecked() != isSelectedInAdapter)
                selectableView.setChecked(isSelectedInAdapter);
            selectableView.setText(getItemName(item));
            selectableView.setIcon(getItemIcon(item), getItemIconBackground(item));
        }

        private void recycle() {
            selectableView.setIcon(null, null);
        }
    }
}

File: lightbulb-recycleradapters/src/main/java/com/github/rooneyandshadows/lightbulb/recycleradapters/implementation/CheckBoxSelectableAdapter.kt
package com.github.rooneyandshadows.lightbulb.recycleradapters.implementation

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView.*
import com.github.rooneyandshadows.lightbulb.recycleradapters.R
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterDataModel
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterSelectableModes.*
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyRecyclerAdapter
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.collection.BasicCollection
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.collection.ExtendedCollection
import com.github.rooneyandshadows.lightbulb.selectableview.CheckBoxView

@Suppress("UNUSED_PARAMETER", "unused", "MemberVisibilityCanBePrivate")
@JvmSuppressWildcards
open class CheckBoxSelectableAdapter<ItemType : EasyAdapterDataModel> : EasyRecyclerAdapter<ExtendedCollection<ItemType>>() {

    override fun createCollection(): ExtendedCollection<ItemType> {
        return ExtendedCollection(this, SELECT_MULTIPLE)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val context = parent.context
        val checkBoxView = LayoutInflater.from(parent.context).inflate(
            R.layout.layout_checkbox_button,
            parent, false
        ) as CheckBoxView
        val padding = getItemPadding(context)
        if (padding != null && padding.size == 4) checkBoxView.setPadding(
            padding[0],
            padding[1],
            padding[2],
            padding[3]
        )
        return CheckBoxViewHolder(checkBoxView)
    }

    @Suppress("UNCHECKED_CAST")
    @Override
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val vHolder: CheckBoxViewHolder = holder as CheckBoxSelectableAdapter<ItemType>.CheckBoxViewHolder
        vHolder.bindItem()
    }

    @Suppress("UNCHECKED_CAST")
    @Override
    override fun onViewRecycled(holder: ViewHolder) {
        val vh: CheckBoxViewHolder = holder as CheckBoxSelectableAdapter<ItemType>.CheckBoxViewHolder
        vh.recycle()
    }

    @Override
    override fun getItemCount(): Int {
        return collection.size()
    }

    protected open fun getItemIcon(context: Context, item: ItemType): Drawable? {
        return null
    }

    protected open fun getItemIconBackground(context: Context, item: ItemType): Drawable? {
        return null
    }

    protected open fun getItemPadding(context: Context): IntArray? {
        return null
    }

    inner class CheckBoxViewHolder internal constructor(checkBoxView: CheckBoxView) : ViewHolder(checkBoxView) {
        private var selectableView: CheckBoxView = itemView as CheckBoxView

        fun bindItem() {
            selectableView.apply {
                val item = collection.getItem(bindingAdapterPosition) ?: return
                val isSelectedInAdapter = collection.isItemSelected(item)
                val itemText = collection.getItemName(item)
                val itemIcon = getItemIcon(context, item)
                val itemIconBackground = getItemIconBackground(context, item)
                if (isChecked != isSelectedInAdapter) isChecked = isSelectedInAdapter
                text = itemText
                setIcon(itemIcon, itemIconBackground)
            }
        }

        fun recycle() {
            selectableView.setIcon(null, null)
        }

        init {
            with(selectableView) {
                setOnCheckedListener { _: CheckBoxView?, isChecked: Boolean ->
                    post {
                        collection.selectItemAt(bindingAdapterPosition, isChecked)
                    }
                }
            }
        }
    }
}